cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -313,54 +275,99 @@ struct rle_stream {
+            decode_index_shared = decode_index;
+          }
+          fill_index_shared = fill_index;
         }
       }
-      __syncthreads();
-
-      // if we haven't run out of space, retrieve the next batch. otherwise leave it for the next
-      // call.
-      if (!t && values_processed < output_count) {
-        thrust::tie(run_start, num_runs) = get_run_batch();
+      // remaining warps decode the runs, starting on the second iteration of this. the pipeline of
+      // runs is also persistent across calls to decode_next, so on the second call to decode_next,
+      // this branch will start doing work immediately.
+      // do/while loop (decode_index == -1 means "first iteration", so we should skip decoding)
+      else if (decode_index >= 0 && decode_index + warp_decode_id < fill_index) {
+        int const run_index = decode_index + warp_decode_id;
+        auto& run           = runs[rolling_index<run_buffer_size>(run_index)];
+        // this is the total amount (absolute) we will write in this invocation
+        // of `decode_next`.
+        int const max_count = cur_values + output_count;
+        // run.output_pos is absolute position, we start decoding
+        // if it's supposed to fit in this call to `decode_next`.
+        if (max_count > run.output_pos) {
+          int remaining        = run.remaining;
+          int const run_offset = run.size - remaining;
+          // last_run_pos is the absolute position of the run, including
+          // what was decoded last time.
+          int const last_run_pos = run.output_pos + run_offset;
+
+          // the amount we should process is the smallest of current remaining, or
+          // space available in the output buffer (for that last run at the end of
+          // a call to decode_next).
+          int const batch_len = min(remaining, max_count - last_run_pos);
+          decode<level_t, max_output_values>(output,
+                                             run.level_run,
+                                             run.start,
+                                             end,
+                                             run.output_pos,
+                                             run_offset,
+                                             batch_len,
+                                             level_bits,
+                                             warp_lane);
+          if (warp_lane == 0) {
+            // after writing this batch, are we at the end of the output buffer?
+            auto const at_end = ((last_run_pos + batch_len - cur_values) == output_count);
+
+            // update remaining for my warp
+            remaining -= batch_len;
+            // this is the last batch we will process this iteration if:
+            // - either this run still has remaining
+            // - or it is consumed fully and its last index corresponds to output_count
+            if (remaining > 0 || at_end) { values_processed_shared = output_count; }
```

Comment:
It will be.  There's only 2 possible cases.  
- If we run out of space in the output, only 1 warp will have a remaining value > 0
 or
- if we're at the very end, only 1 warp will have it's final output position be == to output_count